README.md
# CDAP Cloudera Manager Integration CSD

## CDAP Custom Service Descriptor (CSD)

The CDAP Custom Service Descriptor (CSD) is an extension for Cloudera Manager allowing it to install
and manage CDAP services.  It consists of a service descriptor SDL and a control script, which together
are responsible for configuring and launching CDAP services.

Cloudera's documentation for CSDs can be found [here](https://github.com/cloudera/cm_ext/wiki/CSD-Overview).

Full documentation for installing the CDAP services via Cloudera Manager can be found [here](http://docs.cdap.io/cdap/current/en/integrations/index.html).

### Building the CSD
```
  git clone git@github.com:caskdata/cm_csd.git
  cd cm_csd
  mvn clean package
```

### Installing the CSD

Copy the generated jar file to the CSD directory of a Cloudera Manager instance, by default `/opt/cloudera/csd`.
Then restart Cloudera Manager.

### Installing CDAP via Cloudera Manager

Please refer to the [documentation](http://docs.cdap.io/cdap/current/en/integrations/partners/cloudera/configuring.html) for further CDAP installation instructions.

## License and Trademarks

Copyright © 2014-2015 Cask Data, Inc.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the
License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
either express or implied. See the License for the specific language governing permissions
and limitations under the License.

Cask is a trademark of Cask Data, Inc. All rights reserved.

Apache, Apache HBase, and HBase are trademarks of The Apache Software Foundation. Used with
permission. No endorsement by The Apache Software Foundation is implied by the use of these marks.



CHANGELOG.md
CDAP CSD CHANGELOG
==================

v6.1.1 (Jan 23, 2020)
---------------------
- Add configurations for CDAP 6.1 release ( Issues: #210 [CDAP-16214](https://issues.cask.co/browse/CDAP-16214))

v6.0.0 (May 17, 2019)
---------------------
- Update Market URL ( Issues: #206)
- Package renaming to cdap.io ( Issues: #208)
- Add configurations for CDAP 6.0 release ( Issues: #209 [CDAP-15383](https://issues.cask.co/browse/CDAP-15383))

v5.1.0 (Oct 11, 2018)
---------------------
- Add configurations for CDAP 5.1 release ( Issues: #199 [CDAP-14495](https://issues.cask.co/browse/CDAP-14495) )

v5.0.0 (Jul 30, 2018)
---------------------
- Update default configurations for better usability ( Issues: #187 [CDAP-12767](https://issues.cask.co/browse/CDAP-12767) )
- Update default JVM options for JDK 8 ( Issues: #191 )
- Update default twill JVM options for JDK 8 ( Issues: #192 [CDAP-13416](https://issues.cask.co/browse/CDAP-13416) )
- Add configurations for CDAP 5.0 release ( Issues: #193 #194 #195 [CDAP-13708](https://issues.cask.co/browse/CDAP-13708) )

v4.3.1 (Dec 5, 2017)
--------------------
- Add optional dependency on Sentry ( Issues: #178 )
- Add configurable graceful shutdown timeout ( Issues: #181 #183 [CDAP-12721](https://issues.cask.co/browse/CDAP-12721) )

v4.3.0 (Aug 25, 2017)
---------------------
- Adds compatibility check for CDAP Parcel ( Issues: #172 #173 [CDAP-4874](https://issues.cask.co/browse/CDAP-4874) )
- Adds configurations for CDAP 4.3 release ( Issues: #175 [CDAP-12235](https://issues.cask.co/browse/CDAP-12235) )
- Removes configurations which are designated final ( Issues: #176 [CDAP-11944](https://issues.cask.co/browse/CDAP-11944) )

v4.2.2 (June 13, 2017)
----------------------
- Add cdap-env.sh to gateway role ( Issues: #168 [CDAP-11900](https://issues.cask.co/browse/CDAP-11900) )

v4.2.1 (June 7, 2017)
---------------------
- Update parcel repo ( Issues: #165 )

v4.2.0 (June 6, 2017)
---------------------
- Update start script to handle spark2 ( Issues: #156 [CDAP-11586](https://issues.cask.co/browse/CDAP-11586) )
- Adds configurations for CDAP 4.2 release ( Issues: #157 #159 #160 #162 [CDAP-11531](https://issues.cask.co/browse/CDAP-11531) )
- Adds optional tag & service dependencies on spark 2 ( Issues: #158 [CDAP-11606](https://issues.cask.co/browse/CDAP-11606) )
- Adds JobQueue debug tool ( Issues: #161 [CDAP-11515](https://issues.cask.co/browse/CDAP-11515) )

v4.1.2 (Mar 14, 2017)
---------------------
- Fixes an issue with the log.process.pipeline.lib.dir delimiter ( Issues: #152 [CDAP-8897](https://issues.cask.co/browse/CDAP-8897) )
- Adds optional authentication parameters ( Issues: #153 [CDAP-8891](https://issues.cask.co/browse/CDAP-8891) )

v4.1.1 (Feb 28, 2017)
---------------------
- Update to v2 market url ( Issues: #151 )

v4.1.0 (Feb 27, 2017)
---------------------
- Adds CDAP coprocessor check to Master startup ( Issues: #135, 138, 139 )
- Adds basic rolling restart capability ( Issues #136 [CDAP-5453](https://issues.cask.co/browse/CDAP-5453) )
- Adds configurations for CDAP 4.1 release ( Issues #137 #141 #142 #143 #144 #145 #146 [CDAP-8393](https://issues.cask.co/browse/CDAP-8393) )
- Relaxes required fields ( Issues #138 [CDAP-8435](https://issues.cask.co/browse/CDAP-8435) )

v4.0.1 (Feb 1, 2017)
--------------------
- Exposes configuration for scheduler.misfire.threshold.ms ( Issue: #129 )
- SSL support ( Issues: #130, #131 )

v4.0.0 (Dec 16, 2016)
---------------------
- Adds support for CDAP 4.0 init framework ( Issues: #105 #107 #112 #126 )
- Sync logback-container.xml updates from CDAP ( Issues: #113 #114 #115 )
- Updates UI default port to 11011 ( Issues: #118 [CDAP-5897](https://issues.cask.co/browse/CDAP-5897) )
- Supports improved CDAP mechanism to use Hive Classpath ( Issues #119 [CDAP-7054](https://issues.cask.co/browse/CDAP-7354) )
- Adds capability to run a CDAP debugger utility from the Actions menu ( Issues: #122 [CDAP-5632](https://issues.cask.co/browse/CDAP-5632) )
- Adds CM enhanced JVM support for CDAP services, **Requires CM 5.7 or above** ( Issues: #123 [CDAP-7555](https://issues.cask.co/browse/CDAP-7555) )
- Fixes an issue where CDAP would not pickup latest configuration changes ( Issues: #124 [CDAP-7556](https://issues.cask.co/browse/CDAP-7556) )
- Adds configurations for CDAP 4.0 release ( Issues: #125 #127 [CDAP-7796](https://issues.cask.co/browse/CDAP-7556) )

v3.6.0 (Oct 4, 2016)
--------------------
- Adds configurations for CDAP 3.6 release ( Issues: #108 )

v3.5.1 (Aug 24, 2016)
---------------------
- Fixes an issue where the Upgrade Tool commands could fail ( Issues: #104 )

v3.5.0 (Aug 19, 2016)
---------------------
- Exposes YARN scheduler configurations ( Issues: #96 [CDAP-6182](https://issues.cask.co/browse/CDAP-6182) )
- Fixes an issue where the Router and UI roles needed to be colocated ( Issues: #98 [CDAP-2501](https://issues.cask.co/browse/CDAP-2501) )
- Fixes an issue where the Gateway client was missing router.server.* configurations ( Issues: #98 [CDAP-6974](https://issues.cask.co/browse/CDAP-6974) )
- Adds configurations for CDAP 3.5 release ( Issues: #99 #100 #101 )

v3.4.2 (Aug 4, 2016)
--------------------
- Updates CDAP label and description ( Issues: #90 )
- Prevent unnecessary checks of Hive, HBase, and Spark environments ( Issues: #93 [CDAP-6923](https://issues.cask.co/browse/CDAP-6923) )

v3.4.1 (May 11, 2016)
---------------------
- Fixes an issue where java heap size for the Upgrade commands was not set properly ( Issues: #84 [CDAP-5891](https://issues.cask.co/browse/CDAP-5891) )
- Fixes an issue where the Additional Java Options property was not set properly ( Issues: #86 [CDAP-5956](https://issues.cask.co/browse/CDAP-5956) )

v3.4.0 (Apr 28, 2016)
---------------------
- Increase CDAP PermGen memory ( Issues: #69 #70 )
- Fixes an issue where logging directory was not modifiable ( Issues: #72 [CDAP-5620](https://issues.cask.co/browse/CDAP-5620) )

src/aux/client/cdap-env.sh
# Copyright © 2017 Cask Data, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

# DO NOT EDIT THIS FILE MANUALLY
# Content below is populated by Cloudera Manager using the "CDAP Client
# Advanced Configuration Snippet (Safety Valve) for cdap-conf/cdap-env.sh".
# To modify, update this configuration in Cloudera Manager.

# Example:
# export JAVA_OPTS="${JAVA_OPTS} -Dexample.option=value"



src/scripts/cdap-control.sh
#!/usr/bin/env bash

# Copyright © 2015-2016 Cask Data, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

# Expects argument: [CDAP service]
SERVICE=${1}

# Reads a line from a generated properties file in the format "$host:$key=$value", setting those variables.
function readconf {
  local conf
  IFS=':' read host conf <<< "${1}"
  IFS='=' read key value <<< "${conf}"
}

# Reads the 'peerConfig' generated kafka properties file and formulates the kafka quorum
function generate_kafka_quorum {
  local __seed_brokers=()
  for line in `cat ${KAFKA_PROPERTIES}`; do
    readconf "${line}"
    if [ "${key}" == "kafka.server.port" ]; then
      __seed_brokers+=("${host}:${value}")
    fi
  done
  # Join array
  KAFKA_SEED_BROKERS=${__seed_brokers[@]}
  KAFKA_SEED_BROKERS=${KAFKA_SEED_BROKERS// /,}
}

function substitute_cdap_site_tokens {
  local __cdap_site=${1}
  generate_kafka_quorum
  sed -i -e "s#{{HOSTNAME}}#${HOSTNAME}#" -e "s#{{ZK_QUORUM}}#${ZK_QUORUM}#" \
    -e "s#{{KAFKA_SEED_BROKERS}}#${KAFKA_SEED_BROKERS}#" -e "s#{{LOCAL_DIR}}#${LOCAL_DIR}#" \
    -e "s#{{COMPONENT_HOME}}#${COMPONENT_HOME}#" ${__cdap_site}

  if [[ "${cdap_principal}" != "" ]]; then
    # Kerberos is enabled, update cdap-site.xml keytab and principal settings
    sed -i -e "s#{{CDAP_MASTER_KERBEROS_PRINCIPAL}}#${cdap_principal}#" \
      -e "s#{{CDAP_MASTER_KERBEROS_KEYTAB}}#${CONF_DIR}/cdap.keytab#" \
      ${__cdap_site}
  else
    # Remove cdap-site.xml keytab and principal settings
    sed -i -e "s#{{CDAP_MASTER_KERBEROS_PRINCIPAL}}##" -e "s#{{CDAP_MASTER_KERBEROS_KEYTAB}}##" ${__cdap_site}
  fi
}

# Source CSD build version
[[ -e scripts/build-env.sh ]] && source scripts/build-env.sh
echo "CSD_VERSION: ${CSD_VERSION}"

# Parse CDAP version from CDAP_HOME (exported in CDAP parcel)
CDAP_VERSION=${VERSION:-$(basename ${CDAP_HOME} | cut -d- -f2)}

# Extract CDAP major and minor versions
__cdap_maj_version=$(echo ${CDAP_VERSION} | cut -d. -f1)
__cdap_min_version=$(echo ${CDAP_VERSION} | cut -d. -f2)

# Extract CSD major and minor versions
__csd_maj_version=$(echo ${CSD_VERSION} | cut -d. -f1)
__csd_min_version=$(echo ${CSD_VERSION} | cut -d. -f2)

# CDAP 6.0 renamed the package second-level domain
if [[ ${__cdap_maj_version} -lt 6 ]]; then
  PKG_SLD="co.cask"
else
  PKG_SLD="io.cdap"
fi

# Determine relevant CDAP component paths from sourced parcel variables
case ${SERVICE} in
  (auth-server)
    COMPONENT_HOME=${CDAP_AUTH_SERVER_HOME}
    COMPONENT_CONF_SCRIPT=${CDAP_AUTH_SERVER_CONF_SCRIPT}
    MAIN_CLASS=${PKG_SLD}.cdap.security.runtime.AuthenticationServerMain
    MAIN_CLASS_ARGS=
    JAVA_HEAP_VAR=AUTH_JAVA_HEAPMAX
    AUTH_JAVA_HEAPMAX=${AUTH_JAVA_HEAPMAX:--Xmx1024m}
    EXTRA_CLASSPATH="${EXTRA_CLASSPATH}:/etc/hbase/conf/"
    ;;
  (kafka-server)
    COMPONENT_HOME=${CDAP_KAFKA_SERVER_HOME}
    COMPONENT_CONF_SCRIPT=${CDAP_KAFKA_SERVER_CONF_SCRIPT}
    MAIN_CLASS=${PKG_SLD}.cdap.kafka.run.KafkaServerMain
    MAIN_CLASS_ARGS=
    JAVA_HEAP_VAR=KAFKA_JAVA_HEAPMAX
    KAFKA_JAVA_HEAPMAX=${KAFKA_JAVA_HEAPMAX:--Xmx1024m}
    ;;
  (master)
    COMPONENT_HOME=${CDAP_MASTER_HOME}
    COMPONENT_CONF_SCRIPT=${CDAP_MASTER_CONF_SCRIPT}
    MAIN_CLASS=${PKG_SLD}.cdap.data.runtime.main.MasterServiceMain
    MAIN_CLASS_ARGS="start"
    JAVA_HEAP_VAR=MASTER_JAVA_HEAPMAX
    MASTER_JAVA_HEAPMAX=${MASTER_JAVA_HEAPMAX:--Xmx1024m}
    EXTRA_CLASSPATH="${EXTRA_CLASSPATH}:/etc/hbase/conf/"
    ;;
  (router)
    COMPONENT_HOME=${CDAP_ROUTER_HOME}
    COMPONENT_CONF_SCRIPT=${CDAP_ROUTER_CONF_SCRIPT}
    MAIN_CLASS=${PKG_SLD}.cdap.gateway.router.RouterMain
    MAIN_CLASS_ARGS=
    JAVA_HEAP_VAR=ROUTER_JAVA_HEAPMAX
    ROUTER_JAVA_HEAPMAX=${ROUTER_JAVA_HEAPMAX:--Xmx1024m}
    ;;
  (ui)
    COMPONENT_HOME=${CDAP_UI_HOME}
    COMPONENT_CONF_SCRIPT=${CDAP_UI_CONF_SCRIPT}
    MAIN_CMD=node
    if test -x ${CDAP_HOME}/ui/bin/node ; then
      ${CDAP_HOME}/ui/bin/node --version >/dev/null 2>&1
      if [ $? -eq 0 ] ; then
        MAIN_CMD=${CDAP_HOME}/ui/bin/node
      elif [[ $(which node 2>/dev/null) ]]; then
        MAIN_CMD=node
      else
        echo "Unable to locate Node.js binary (node), is it installed and in the PATH?"
        exit 1
      fi
    fi
    export NODE_ENV=production
    # UI Entry point renamed in 6.0
    if [[ ${__cdap_maj_version} -lt 6 ]]; then
      MAIN_CMD_ARGS="${CDAP_HOME}/ui/server.js"
    else
      MAIN_CMD_ARGS="${CDAP_HOME}/ui/index.js"
    fi
    ;;
  (client)
    CLIENT_CONF_DIR=${CONF_DIR}/cdap-conf
    HOSTNAME=`hostname -f`
    substitute_cdap_site_tokens ${CLIENT_CONF_DIR}/cdap-site.xml
    exit 0
    ;;
  (upgrade_hbase)
    # The upgrade tool is run as master, but with an overridden $MAIN_CLASS and $MAIN_CLASS_ARGS
    COMPONENT_HOME=${CDAP_MASTER_HOME}
    MAIN_CLASS=${PKG_SLD}.cdap.data.tools.UpgradeTool
    MAIN_CLASS_ARGS="upgrade_hbase force"
    # Set heap max, normally set in COMPONENT_CONF_SCRIPT
    JAVA_HEAPMAX=${MASTER_JAVA_HEAPMAX:--Xmx1024m}
    ;;
  (upgrade)
    # The upgrade tool is run as master, but with an overridden $MAIN_CLASS and $MAIN_CLASS_ARGS
    COMPONENT_HOME=${CDAP_MASTER_HOME}
    MAIN_CLASS=${PKG_SLD}.cdap.data.tools.UpgradeTool
    MAIN_CLASS_ARGS="upgrade force"
    # Set heap max, normally set in COMPONENT_CONF_SCRIPT
    JAVA_HEAPMAX=${MASTER_JAVA_HEAPMAX:--Xmx1024m}
    ;;
  (postupgrade)
    # A post-upgrade step to correct any pending flow metrics. Kafka server must be running
    COMPONENT_HOME=${CDAP_MASTER_HOME}
    MAIN_CLASS=${PKG_SLD}.cdap.data.tools.flow.FlowQueuePendingCorrector
    MAIN_CLASS_ARGS=""
    # Set heap max, normally set in COMPONENT_CONF_SCRIPT
    JAVA_HEAPMAX=${MASTER_JAVA_HEAPMAX:--Xmx1024m}
    ;;
  (debugger_utility)
    # The debugger utility (ie HBase queue debugger) is run as master, but with an overridden $MAIN_CLASS
    COMPONENT_HOME=${CDAP_MASTER_HOME}
    # MAIN_CLASS set by CSD, configurable by user
    # MAIN_CLASS_ARGS set by CSD, configurable by user
    # Set heap max, normally set in COMPONENT_CONF_SCRIPT
    JAVA_HEAPMAX=${MASTER_JAVA_HEAPMAX:--Xmx1024m}
    ;;
  (*)
    echo "Unknown service specified: ${SERVICE}"
    exit 1
    ;;
esac

# Source Cloudera common functions (for Kerberos)
echo "Sourcing Cloudera Common Script: ${COMMON_SCRIPT}"
source ${COMMON_SCRIPT}

# Source the CDAP common init functions
if [[ -e ${COMPONENT_HOME}/bin/functions.sh ]]; then
  echo "Sourcing CDAP common functions: ${COMPONENT_HOME}/bin/functions.sh"
  source ${COMPONENT_HOME}/bin/functions.sh
else
  echo "Sourcing CDAP pre-4.0 common functions: ${COMPONENT_HOME}/bin/common[-env].sh"
  source ${COMPONENT_HOME}/bin/common-env.sh
  source ${COMPONENT_HOME}/bin/common.sh
fi

# Remap CDAP common init functions, if necessary
fn_exists() { type -t ${1} | grep -q function; }

fn_exists cdap_set_classpath || cdap_set_classpath() { set_classpath ${*}; }
fn_exists cdap_set_hive_classpath || cdap_set_hive_classpath() { set_hive_classpath ${*}; }
fn_exists cdap_set_hbase || cdap_set_hbase() { set_hbase ${*}; }

# Redefine cdap_kinit to be a no-op. CM handles kinit for us
cdap_kinit() { return 0; }

# Token replacement in CM-generated cdap-site.xml
# Hostname
HOSTNAME=`hostname -f`
substitute_cdap_site_tokens ${CONF_DIR}/cdap-site.xml

# Copy logback-container.xml into place unless user has populated safety valve
if [[ -s logback-container.xml ]]; then
  echo "Populating logback-container.xml from safety valve content. Ensure the contents of the safety valve represent the entire file!"
else
  # No safety valve content, copy the packaged default into place
  cp aux/logback-container.xml.default logback-container.xml
fi

# Source CDAP Component config if defined and readable
if [[ -r "${COMPONENT_CONF_SCRIPT}" ]]; then
  source ${COMPONENT_CONF_SCRIPT}
fi

# Set JAVA_HEAPMAX from variable defined in JAVA_HEAP_VAR, unless defined already
JAVA_HEAPMAX=${JAVA_HEAPMAX:-${!JAVA_HEAP_VAR}}
export JAVA_HEAPMAX

# CDAP_CONF is used by Web-App to find cdap-site.xml
export CDAP_CONF=${CONF_DIR}

if [ "${cdap_principal}" != "" ]; then
  # Runs kinit
  export SCM_KERBEROS_PRINCIPAL=${cdap_principal}
  acquire_kerberos_tgt cdap.keytab
fi

# Check compatibility between CSD and Parcel between maj.min versions
if [[ "${CSD_COMPATIBILITY_CHECK_ENABLED}" == "true" ]]; then
  compare_versions ${__csd_maj_version}.${__csd_min_version} ${__cdap_maj_version}.${__cdap_min_version}
  __ret=$?
  if [[ ${__ret} -gt 0 ]]; then
    # CDAP Parcel version is greater than CSD
    echo "ERROR: Detected Parcel version ${CDAP_VERSION} is newer than CSD version ${CSD_VERSION}. Please upgrade the CDAP CSD. This check can be disabled via the csd.compatibility.check.enabled configuration in Cloudera Manager."
    exit 1
  else
    echo "CSD/Parcel compatibility confirmed"
  fi
else
  echo "CSD/Parcel compatibility check disabled"
fi

# Debug info
echo "CDAP_HOME: ${CDAP_HOME}"
echo "CDAP_VERSION: ${CDAP_VERSION}"
echo "COMPONENT_HOME: ${COMPONENT_HOME}"
echo "COMPONENT_CONF_SCRIPT: ${COMPONENT_CONF_SCRIPT}"
echo "CONF_DIR: ${CONF_DIR}"
echo "ENV: `env`"

# Launch a cmd or a java app
if [ ${MAIN_CLASS} ]; then
  # Launch a java app

  # Set java
  JAVA=${JAVA_HOME}/bin/java

  # Set HBASE_HOME to CM-provided active location
  export HBASE_HOME=${CDH_HBASE_HOME}

  # CDAP-7556: we must construct the CLASSPATH to include our dependencies' configuration first
  # Dependency service configs will be in ${CONF_DIR}/[service]-conf/ directories
  for d in $(ls -1d ${CONF_DIR}/*-conf/); do
    if [[ "$d" == *spark-conf/ ]] && [[ ${SPARK_MAJOR_VERSION} -eq 2 ]]; then
      continue
    fi
    if [[ "$d" == *spark2-conf/ ]] && [[ ${SPARK_MAJOR_VERSION} -ne 2 ]]; then
      echo "WARNING: Spark 2 dependency detected but SPARK_MAJOR_VERSION not set to 2. Defaulting to Spark 1. Please set SPARK_MAJOR_VERSION=2 in the Service Environment Safety Valve to use Spark 2"
      continue
    fi
    if [[ -n ${CLASSPATH} ]]; then
      CLASSPATH="${CLASSPATH}:${d}"
    else
      CLASSPATH=${d}
    fi
  done

  # Set base classpath to include component and conf directory (CM provided)
  cdap_set_classpath ${COMPONENT_HOME} ${CONF_DIR}

  # Prepend CM-provided ${CSD_JAVA_OPTS} to our ${CDAP_JAVA_OPTS}
  if [[ -n ${CSD_JAVA_OPTS} ]]; then
    CDAP_JAVA_OPTS="${CSD_JAVA_OPTS} ${CDAP_JAVA_OPTS}"
  fi

  echo "`date` Starting Java service ${SERVICE} on `hostname`"
  "${JAVA}" -version
  echo "`ulimit -a`"
  echo "Using java_heapmax: ${JAVA_HEAPMAX}"
  echo "Using user.dir: ${LOCAL_DIR}"
  echo "Using classpath: ${CLASSPATH}"
  echo "Using main_class: ${MAIN_CLASS}"
  echo "Using args: ${MAIN_CLASS_ARGS}"

  # Run Master-specific logic (for master and upgrade* services)
  if [ ${COMPONENT_HOME} == ${CDAP_MASTER_HOME} ]; then

    # Set HIVE_HOME to CM-provided active location
    export HIVE_HOME=${CDH_HIVE_HOME}
    # Setup hive classpath if hive is installed, this has to be run after HBASE_CP is setup by set_classpath.
    cdap_set_hive_classpath

    # Include appropriate hbase_compat module in classpath
    cdap_set_hbase

    # if no spark is available, compat is still spark1
    export SPARK_COMPAT="spark1_2.10"
    # If a user has selected a dependency on Spark, CM will generate spark-conf directory in CWD
    # spark-conf may also be there if Hive depends on spark and CDAP depends on Hive
    if [ -d "spark2-conf" ] && [[ ${SPARK_MAJOR_VERSION} -eq 2 ]]; then
      export SPARK_HOME=${CDH_SPARK2_HOME}
      export SPARK_COMPAT="spark2_2.11"
    elif [ -d "spark-conf" ]; then
      export SPARK_HOME=${CDH_SPARK_HOME}
    fi

    # CDAP < 4.0 uses explore.conf.files, CDAP >= 4.0 uses explore.conf.dirs
    if [[ -n ${EXPLORE_CONF_DIRS} ]]; then
      __EXPLORE_CONF_FLAG="-Dexplore.conf.dirs=${EXPLORE_CONF_DIRS}"
    else
      __EXPLORE_CONF_FLAG="-Dexplore.conf.files=${EXPLORE_CONF_FILES}"
    fi
    echo "Using ${__EXPLORE_CONF_FLAG}"
    echo "Using explore.classpath: ${EXPLORE_CLASSPATH}"

    echo "Using SPARK_HOME: ${SPARK_HOME}"

    # Build and upload coprocessor jars, if we are starting master
    if [[ "${MAIN_CLASS}" == "${PKG_SLD}.cdap.data.runtime.main.MasterServiceMain" ]]; then
      # Run only if CDAP_VERSION >= 4.1
      if [[ __cdap_maj_version -gt 4 ]] || [[ __cdap_maj_version -ge 4 && __cdap_min_version -ge 1 ]]; then
        echo "$(date) Ensuring required HBase coprocessors are on HDFS"
        cdap_setup_coprocessors </dev/null 2>&1 || die "Could not setup coprocessors."
      fi
    fi

    if [[ "${STARTUP_CHECKS_ENABLED}" == "true" ]]; then
      # Run only if CDAP_VERSION >= 3.3
      if [[ __cdap_maj_version -gt 3 ]] || [[ __cdap_maj_version -ge 3 && __cdap_min_version -ge 3 ]]; then
        echo "Running startup checks -- this may take a few minutes"
        echo "Checks can be disabled using the master.startup.checks.enabled configuration option"
        "${JAVA}" "${JAVA_HEAPMAX}" \
          ${__EXPLORE_CONF_FLAG} \
          -Dexplore.classpath=${EXPLORE_CLASSPATH} ${CDAP_JAVA_OPTS} \
          -Dcdap.home=${CDAP_HOME} \
          -cp ${CLASSPATH} \
          ${PKG_SLD}.cdap.master.startup.MasterStartupTool 2>&1
        if [ $? -ne 0 ]; then
          echo "Master startup checks failed. Please check the CDAP Master Role logs to address issues"
          exit 1
        fi
      else
        echo "Skipping Master startup checks as CDAP version ${CDAP_VERSION} does not support them"
      fi
    fi
  fi

  # Exec into Master Service
  exec "${JAVA}" -Dcdap.service=${SERVICE} "${JAVA_HEAPMAX}" \
    ${__EXPLORE_CONF_FLAG} \
    -Dexplore.classpath=${EXPLORE_CLASSPATH} ${CDAP_JAVA_OPTS} \
    -Duser.dir=${LOCAL_DIR} \
    -Dcdap.home=${CDAP_HOME} \
    -cp ${CLASSPATH} ${MAIN_CLASS} ${MAIN_CLASS_ARGS}

elif [ ${MAIN_CMD} ]; then
  # Launch a non-java app

  echo "`date` Starting service ${SERVICE} on ${HOSTNAME}"
  echo "`ulimit -a`"

  exec ${MAIN_CMD} ${MAIN_CMD_ARGS}
fi
